lib/awful/widget/layoutlist.lua

```diff
@@ -0,0 +1,151 @@
+----------------------------------------------------------------------------
+--- Display the available client layouts for a screen.
+--
+--@DOC_awful_widget_layoutlist_popup_EXAMPLE@
+--
+--@DOC_awful_widget_layoutlist_bar_EXAMPLE@
+--
+-- @author Emmanuel Lepage Vallee &lt;elv1313@gmail.com&gt;
+-- @copyright 2017 Emmanuel Lepage Vallee
+-- @classmod awful.widget.layoutlist
+----------------------------------------------------------------------------
+
+local wibox    = require("wibox")
+local awcommon = require("awful.widget.common")
+local abutton  = require("awful.button")
+local gtable   = require("gears.table")
+local beautiful= require("beautiful")
+local alayout  = require("awful.layout")
+local surface  = require("gears.surface")
+
+local module = {}
+
+local default_buttons = gtable.join(
+    abutton({ }, 1, function(a) a.callback() end)
+)
+
+local function wb_label(item)
+    local selected = alayout.get(item.screen) == item.layout
+    return item.layout.name, selected and beautiful.bg_focus or nil, nil, item.icon
+end
+
+local function reload_cache(self)
+    self._private.cache = {}
+
+    for _, l in ipairs(alayout.layouts or {}) do
+        local icn_path, icon = beautiful["layout_" .. (l.name or "")]
+        if icn_path then
+            icon = surface.load(icn_path)
+        end
+
+        table.insert(self._private.cache, {
+            icon     = icon,
+            layout   = l,
+            screen   = self.screen,
+            callback = function() end,
+        })
+    end
+end
+
+local function update(self)
+    if not self._private.layout then return end
+    awcommon.list_update(
+        self._private.layout,
+        default_buttons,
+        wb_label,
+        self._private.data,
+        self._private.cache,
```

Comment:
Why is the cache necessary here? `gears.surface.load` has its own cache and besides that `reload_cache` doesn't do anything expensive. Just use `alayout.layouts` directly here and do "the necessary stuff" in `wb_label`.

Reply:
The other `awful.widget.common` widgets work like that. I could be changed later on.